enhancements/ingress/logging-api.md

```diff
@@ -0,0 +1,401 @@
+
+	// Syslog sends log messages to a syslog endpoint.
+	SyslogLoggingDestinationType LoggingDestinationType = "Syslog"
+
+	// ContainerLoggingSidecarContainerName is the name of the container
+	// with the log output in an ingress controller pod when container
+	// logging is used.
+	ContainerLoggingSidecarContainerName = "logs"
+)
+```
+
+For `Syslog`, an endpoint must be specified, and a facility may be specified (if
+unspecified the default is "local1"); `Container` has no parameters:
+
+```go
+// SyslogLoggingDestinationParameters describes parameters for the Syslog
+// logging destination type.
+type SyslogLoggingDestinationParameters struct {
+	// endpoint identifies the syslog endpoint that receives log messages.
+	// The endpoint must be a host name or IP address and a UDP port number.
+	// Specify the value in the format of a colon-separated name or address
+	// and port number: host:port
+	//
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:MinLength=1
+	// +required
+	Endpoint string `json:"endpoint"`
+
+	// facility specifies the syslog facility of log messages.
+	//
+	// If this field is empty, the facility is "local1".
+	//
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:validation:Enum=kern;user;mail;daemon;auth;syslog;lpr;news;uucp;cron;auth2;ftp;ntp;audit;alert;cron2;local0;local1;local2;local3;local4;local5;local6;local7
+	// +optional
+	Facility string `json:"facility,omitempty"`
+}
+
+// ContainerLoggingDestinationParameters describes parameters for the Container
+// logging destination type.
+type ContainerLoggingDestinationParameters struct {
+}
+```
+
+To disable access logging, the user leaves `spec.logging` or
+`spec.logging.access` unspecified.  Otherwise the user must specify a
+destination using `spec.logging.access.destination`.  To specify a destination,
+the user must specify either `Container` or `Syslog` for
+`spec.logging.access.destination.type`.  If the destination type is `Syslog`,
+the user must specify a destination endpoint using
+`spec.logging.access.destination.syslog.endpoint` and may specify a facility
+using `spec.logging.access.destination.syslog.facility`.  The user may specify
+`spec.logging.access.httpLogFormat` to customize the log format.  For example, the following is the definition of an IngressController that logs to a syslog endpoint with IP address 1.2.3.4 and port 10514:
+
+```yaml
+apiVersion: operator.openshift.io/v1
+kind: IngressController
+metadata:
+  name: default
+  namespace: openshift-ingress-operator
+spec:
+  replicas: 2
+  endpointPublishingStrategy:
+    type: NodePortService
+  logging:
+    access:
+      destination:
+        type: Syslog
+        syslog:
+          endpoint: 1.2.3.4:10514
+```
+
+#### Validation
+
+Omitting `spec.logging` and omitting `spec.logging.access` have well defined
+semantics.
+
+The `spec.logging.access.httpLogFormat` field value is a free-form format
+string, so it is not validated.
```

Comment:
I still have concerns about treating this as opaque. What should happen if the implementation can't deal with the provided format?

And in our specific case (openshift/router), I suspect passing garbage all the way through environment could crash haproxy. So maybe the ingresscontroller would go degraded. Would that be the expectation to communicate (that invalid formats can cause the ingresscontroller to be unavailable/degraded)?

Reply:
Is this something that we can follow up on? Not sure if it should block the EP